crates/consensus/beacon/src/engine/hooks/mod.rs

```diff
@@ -0,0 +1,116 @@
+use reth_interfaces::sync::SyncState;
+use reth_primitives::BlockNumber;
+use std::task::{Context, Poll};
+
+mod controller;
+pub(crate) use controller::HooksController;
+
+mod prune;
+pub use prune::PruneHook;
+
+/// Collection of [hooks][`Hook`].
+#[derive(Default)]
+pub struct Hooks {
+    inner: Vec<Box<dyn Hook>>,
+}
+
+impl Hooks {
+    /// Creates a new empty collection of [hooks][`Hook`].
+    pub fn new() -> Self {
+        Self { inner: Vec::new() }
+    }
+
+    /// Adds a new [hook][`Hook`] to the collection.
+    pub fn add<H: Hook>(&mut self, hook: H) {
+        self.inner.push(Box::new(hook))
+    }
+}
+
+/// Hook that will be run during the main loop of
+/// [consensus engine][`crate::engine::BeaconConsensusEngine`].
+pub trait Hook: Send + Sync + 'static {
+    /// Returns a human-readable name for the hook.
+    fn name(&self) -> &'static str;
+
+    /// Advances the hook execution, emitting an [event][`HookEvent`].
+    fn poll(&mut self, cx: &mut Context<'_>, args: HookArguments) -> Poll<HookEvent>;
+
+    /// Returns an [action][`HookAction`], if any, according to the passed [event][`HookEvent`].
+    fn on_event(&mut self, event: HookEvent) -> Result<Option<HookAction>, HookError>;
```

Comment:
this I don't understand after reading the event docs.

why do we feed the outcome of poll back into an event callback, this looks unnecessary because the hook already knows this.

Reply:
we act on `Started/Finished` here: https://github.com/paradigmxyz/reth/blob/2a6fd5c0b4367a2bd808cbc3276ce10a6695701c/crates/consensus/beacon/src/engine/hooks/controller.rs#L64-L68 https://github.com/paradigmxyz/reth/blob/2a6fd5c0b4367a2bd808cbc3276ce10a6695701c/crates/consensus/beacon/src/engine/hooks/controller.rs#L123-L127

also, there might be no action required (as will be for snapshots, for example), but the hook still need to signal as about its status.

We can combine it into one `HookEvent::Finished(HookAction)` though, and it will be returned on hook polling.

Reply:
ok, I just made the `Hook::poll` return both event and an optional action, so we could:
1. Act on event and set/unset the `running_hook_with_db_write`
2. Pass the action further down